### PR TITLE
Add option to show last line with syntax-aware code folding

### PR DESCRIPTION
**Pull request by RussPitcher**
## PR Summary

Add a settings option to enable the last line of a syntax-aware folded section to be displayed when the section is folded. Relates to issue #1514

I have added a speculative note in the changelog.md file as I wasn't sure if I should update it or not.

## PR Checklist

Note: Tick the boxes below that apply to this pull request by putting an `x` between the square brackets.
Please mark anything not applicable to this PR `NA`.

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] This PR is ready to merge and is not work in progress
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready

### CODE REVIEW

**Comment by rjmholt** [comment]
Tagging @glennsarti for review (won't let me add you in the reviewers dropdown...)

**Comment by rjmholt** [comment]
No need to add to the changelog -- even though I would like it to be done in each PR, it would create merge conflicts for *every* PR. So we generate it at release time. Just make sure you include a pithy description in your commit messages (which we squash, but will do our best to preserve information from).

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,8 @@
 # vscode-powershell Release History
 
+- [vscode-PowerShell #????](https://github.com/PowerShell/vscode-PowerShell/pull/????) -
```
Comment (rjmholt) [inline CHANGELOG.md]: This will cause merge conflicts later on if we leave it in. If you put your summary from here into a commit message, it will make its way into the changelog.

Just stick the summary into a commit taking out the changelog entry is probably the easiest way to go. Otherwise there is also the `git commit --amend` -> `git push origin +branch` option.

File: package.json [hunk 2]
```diff
@@ -478,6 +478,11 @@
           "default": true,
           "description": "Enables syntax based code folding. When disabled, the default indentation based code folding is used."
         },
+        "powershell.codeFolding.showLastLine": {
+          "type": "boolean",
+          "default": false,
```
Comment (rjmholt) [inline package.json]: @tylerl0706 @rkeithhill @glennsarti @RussPitcher Thoughts on this being the default? (Note as in the description below that this is VSCode's default)

File: src/features/Folding.ts [hunk 3]
```diff
@@ -101,10 +101,13 @@ class LineNumberRange {
      */
     public rangeKind: vscode.FoldingRangeKind;
 
+    private settings: Settings.ISettings;
+
     constructor(
         rangeKind: vscode.FoldingRangeKind,
     ) {
         this.rangeKind = rangeKind;
+        this.settings = Settings.load();
```
Comment (rjmholt) [inline src/features/Folding.ts]: Is there a way to ensure that settings are only loaded once? Like passing them into the constructor? Or does `Settings.load()` guarantee that (I assume it does actually, but want to check)?

File: src/features/Folding.ts [hunk 4]
```diff
@@ -121,6 +124,9 @@ class LineNumberRange {
     ): LineNumberRange {
         this.startline = document.positionAt(start.startIndex).line;
         this.endline = document.positionAt(end.startIndex).line;
+        if (this.settings.codeFolding && this.settings.codeFolding.showLastLine) {
+            this.endline--;
```
Comment (rjmholt) [inline src/features/Folding.ts]: I assume `endline-1` is never going to be a bad value for VSCode. We might want to test some edge cases here.

Comment (glennsarti) [reply src/features/Folding.ts]: Acutally, in retrospect.  This is wrong place to do this.  Because, changing the endline too early means the `isValidRange` function will give out false positives (https://github.com/PowerShell/vscode-powershell/blob/7afec44453836b9b9b54b0505e7ba2857d9b0f84/src/features/Folding.ts#L153)

Instead it should replace https://github.com/PowerShell/vscode-powershell/blob/7afec44453836b9b9b54b0505e7ba2857d9b0f84/src/features/Folding.ts#L160-L162

``` typescript
    public toFoldingRange(settings:ISettings): vscode.FoldingRange {
        if (settings.codeFolding && settings.codeFolding.showLastLine) {
            return new vscode.FoldingRange(this.startline, this.endline - 1, this.rangeKind);
        } else {
            return new vscode.FoldingRange(this.startline, this.endline, this.rangeKind);
        }
    }
```

And then change the `provideFoldingRanges` method to;

https://github.com/PowerShell/vscode-powershell/blob/7afec44453836b9b9b54b0505e7ba2857d9b0f84/src/features/Folding.ts#L255-L261

``` typescript
        const settings = this.currentSettings();
        return foldableRegions
            // It's possible to have duplicate or overlapping ranges, that is, regions which have the same starting
            // line number as the previous region. Therefore only emit ranges which have a different starting line
            // than the previous range.
            .filter((item, index, src) => index === 0 || item.startline !== src[index - 1].startline)
            // Convert the internal representation into the VSCode expected type
            .map((item) => item.toFoldingRange(settings));
    }

    private currentSettings(): ISettings { Settings.load(); }
```

The private method `currentSettings()` is needed so we can mock it in the unit tests to flick the behaviour.

**Comment by glennsarti** [comment]
@rjmholt @RussPitcher I've raised an alternate implementation of this PR at #1551.  I've kept the commit messages horrible but if we choose to use that they can be cleaned up.

**Comment by tylerl0706** [comment]
Closing this in favor of #1557 🙂
